[DC-1016] Map access request to asset definition on Snapshot Create by Request Id

snf2ye opened the pull request:
https://broadworkbench.atlassian.net/browse/DC-1016
____________________
### The Changes

- Remove references to the pre-created asset ("concept_asset")
- Build an Asset based on the access request: in addition to the requested occurrence table, include the person and concept tables as well as the relationships between the tables. 
- Put the asset in the working map to be used a later step. 
- Manually map between domain names and the dataset table (to be remove in DC-1120).
- Added flag on asset to force strict relationship walking on snapshot create. Without this flag, a FROM->TO relationship can be swapped to a TO->FROM relationship. This doesn't work for our snapshot create method.

### Remaining/Follow-on Changes

- [ ] Manually update snapshot builder settings per environment: After an initial review of these changes, I'm going to update snapshot builder settings across our environments to match the example in this code: Removing the placeholder column names and the genomic domain (since it doesn't map to a table in the current dataset). 
- [ ] Add FROM->TO relationships to OMOP datasets to capture the person->occurrence relationships. Currently there are only occurrence->person table relationships.
- [ ] Rework snapshot builder settings to populate the new table model and remove the manual definition of mapping the domain name to the actual dataset table name. (To be done in DC-1120)

### Testing

- Expect the existing integration tests to test whether the snapshot request by Id is nominally correctly created
- Changes are deployed to a BEE (https://data.shelbee.bee.envs-terra.bio/) and used new snapshot create method and snapshot builder settings to create a request and snapshot. 
      - If you want to test out the created snapshots, register here: https://terra.shelbee.bee.envs-terra.bio/ and send me your email. I will add you to the following snapshots/datasets:
      - [GCP Source snapshot](https://data.shelbee.bee.envs-terra.bio/snapshots/995cc0f1-a6de-4475-b149-634b38803845)
      - [GCP Snapshot By Request - With Drug, Procedure, and Condition Value sets](https://data.shelbee.bee.envs-terra.bio/snapshots/48d4b83b-ae5f-4df0-a5fe-5035ab4d4f66)
      - [GCP Snapshot By Request - With Procedure and Drug value sets](https://data.shelbee.bee.envs-terra.bio/snapshots/7a6ce3c3-7b7a-47e9-a0f0-301d0960a942)


### Follow-on Work
1) DC-1120: Update snapshot builder settings
2) [DC-1160](https://broadworkbench.atlassian.net/browse/DC-1160): Fix Azure Snapshot Create - If there is more than one relationship between tables, snapshot create does not correctly select the right tabular data
3) [DC-1161](https://broadworkbench.atlassian.net/browse/DC-1161) - When viewing the snapshot in the TDR UI, the column ordering is different than in the source dataset. It just is particularly noticeable that the person_id tends to be one of the last columns for each occurrence table. 


[DC-1160]: https://broadworkbench.atlassian.net/browse/DC-1160?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

snf2ye committed:
Basic test and use case working

manually populate relationships

snf2ye committed:
remove notes/cleanup

some cleanup

snf2ye committed:
support columns

snf2ye committed:
Missing asset specification from step that builds snapshot table

Fix tests

snf2ye committed:
Remove columns and genomics from snapshot builder settings: We need these to map to real values now

remove values

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,32 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
-        createSnapshotTablesFromDatasetAsset(datasetName, ASSET_NAME, snapshotSource, snapshot);
+        Dataset queryDataset = datasetService.retrieveByName(datasetName);
+        UUID accessRequestId =
+            snapshotRequestModel.getContents().get(0).getRequestIdSpec().getSnapshotRequestId();
```
Since the `switch` that contains this `case` already uses `snapshotRequestModel.getContents().get(0)`, you could get that value outside the switch and re-use it here, now that it's used three times in this path.

Actually, looking at this again, the code is already extracting `requestContents` on line 591 above. It would be nice to replace all the `get(0)` references with that variable.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,32 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+            buildAssetFromSnapshotAccessRequest(queryDataset, accessRequest);
+        // Make sure the AssetSpecification ID field is null
+        // so that it is not written to the database in the snapshot source table
+        queryAssetSpecification.id(null);
```
This API has been around a while, but I think it's confusing to use a fluent-style setting in a non-fluent context. For openapi-generated models, we get both versions, but this is a hand written model object, and doesn't have the setter APIs. I'd add it and use it here, but it's up to you.

snf2ye replied:
Ah, ok. I ended up removing this call.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,32 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+
+        AssetSpecification queryAssetSpecification =
+            buildAssetFromSnapshotAccessRequest(queryDataset, accessRequest);
+        // Make sure the AssetSpecification ID field is null
+        // so that it is not written to the database in the snapshot source table
```
Where is the null check that enforces this? Does `buildAssetFromSnapshotAccessRequest()` return an specification with a non-null ID, and if so, where does it get it from?

snf2ye replied:
buildAssetFromSnapshotAccessRequest() returns a null id. So, this should never be populated. My thought was that this was more for documentation proposes rather than a real check. I ended up removing it, but glad to revisit it.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+  AssetSpecification buildAssetFromSnapshotAccessRequest(
+      Dataset dataset, SnapshotAccessRequestResponse snapshotRequestModel) {
+    // build asset model from snapshot request
+    AssetModel assetModel = new AssetModel();
+    assetModel.name("snapshot-by-request-asset");
+    assetModel.rootTable("person");
+    assetModel.rootColumn("person_id");
```
I think it's clearer to write this using fluent-style syntax
```suggestion
    AssetModel assetModel = new AssetModel()
        .name("snapshot-by-request-asset")
        .rootTable("person")
        .rootColumn("person_id");
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+  @VisibleForTesting
+  List<SnapshotBuilderTable> pullTables(SnapshotAccessRequestResponse snapshotRequestModel) {
+    var valueSets = snapshotRequestModel.getSnapshotSpecification().getValueSets();
+    var valueSetNames = valueSets.stream().map(SnapshotBuilderFeatureValueGroup::getName).toList();
+
+    Map<String, SnapshotBuilderTable> tableMap = populateManualTableMap();
+
+    return valueSetNames.stream()
```
Instead of streaming and collecting `valueSetNames`, you could stream it directly into the `map()`
```suggestion
    var tableMap = populateManualTableMap();

    return valueSets.stream()
        .map(SnapshotBuilderFeatureValueGroup::getName)
```

Although one reason to keep it separate is to generate all invalid names at once.
```
    Set<String> missing = new HashSet<>(valueSetNames);
    missing.removeAll(tableMap.keySet());
    if (!missing.isEmpty()) {
      throw new IllegalArgumentException("Unknown value set names: " + missing);
    }

    return valueSetNames.stream()
        .map(tableMap::get)
        .toList();
```

snf2ye replied:
Oh nice, that rework is better!

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+        .toList();
+  }
+
+  private Map<String, SnapshotBuilderTable> populateManualTableMap() {
```
I wonder if this should be in a separate class so it's clear that it's axin specific?

I guess it won't be around too long so maybe it's OK to leave it here.

rjohanek replied:
Is this AXIN specific or just OMOP specific?

rjohanek replied:
So just to clarify again, all of the information from this map will be stored in the settings for each snapshot so this method will go away entirely and asset creation will always match the configuration of the snapshot?

snf2ye replied:
That's correct - this method will go away with DC-1120.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotByRequestIdInterface.java:
```diff
@@ -1,13 +1,11 @@
 import bio.terra.common.iam.AuthenticatedUserRequest;
-import bio.terra.model.SnapshotAccessRequestResponse;
-import bio.terra.model.SnapshotRequestModel;
+import bio.terra.model.*;
```
Can you expand this wildcard import?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotByRequestIdInterface.java:
```diff
@@ -25,15 +23,17 @@ default StepResult prepareAndCreateSnapshot(
+
     AssetSpecification assetSpecification =
-        SnapshotService.getAssetByNameFromDataset(dataset, SnapshotService.ASSET_NAME);
+        workingMap.get(
+            SnapshotWorkingMapKeys.SNAPSHOT_ASSET_SPECIFICATION, AssetSpecification.class);
```
Instead of storing it in the map, could we re-generate the asset spec here by calling `SnapshotService.buildAssetFromSnapshotAccessRequest()`?

Is there a need to generate the asset spec in `CreateSnapshotMetadataStep`, or can it be moved to this step instead?

snf2ye replied:
We need the asset spec in the CreateSnapshotMetadataStep in order to select which tables are included in the snapshot.
As suggested, I've switched to regenerating the asset spec in both steps by calling that snapshot service method. I couldn't put it in the map because of an infinite recursion error on attempting to put the asset specification into the flight map.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotMetadataStep.java:
```diff
@@ -45,14 +46,23 @@ public CreateSnapshotMetadataStep(
+          .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
+        workingMap.put(
+            SnapshotWorkingMapKeys.SNAPSHOT_ASSET_SPECIFICATION,
+            snapshot.getFirstSnapshotSource().getAssetSpecification());
```
I think trying to make the by request path match up with the regular path makes this part more confusing; the code generates an asset spec and adds it to the snapshot source, but it doesn't store it in the snapshot in the database. Do we need to generate the assset spec here at all, if the mode is `by request`?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotMetadataStep.java:
```diff
@@ -45,14 +46,23 @@ public CreateSnapshotMetadataStep(
+          .getContents()
+          .get(0)
+          .getMode()
+          .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
```
When comparing enums, I think it's better to use `==` instead of `equals()`, as enum instances are final statics.
```suggestion
          == SnapshotRequestContentsModel.ModeEnum.BYREQUESTID) {
```

pshapiro4broad commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -7406,6 +7406,25 @@ components:
           items:
             type: string
 
+    SnapshotBuilderTable:
```
Is the idea that this would eventually be added to `SnapshotBuilderSettings`?

snf2ye replied:
That's correct

snf2ye replied:
I've detailed the future plan in this ticket: https://broadworkbench.atlassian.net/browse/DC-1120

pshapiro4broad commented on src/test/java/bio/terra/service/snapshot/SnapshotServiceTest.java:
```diff
@@ -1327,19 +1329,40 @@ void makeSnapshotFromSnapshotRequestByRequestId() {
+    SnapshotSource snapshotSource = new SnapshotSource().dataset(dataset);
 
-    assertThat(actual.getTables().size(), is(0));
+    assertThat(actual.getTables().size(), is(3));
```
When validating the size of a collection, you can use `hasSize()`
```suggestion
    assertThat(actual.getTables(), hasSize(3));
```

pshapiro4broad reviewed (commented):
Some comments so far

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,32 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+        AssetSpecification queryAssetSpecification =
+            buildAssetFromSnapshotAccessRequest(queryDataset, accessRequest);
+        // Make sure the AssetSpecification ID field is null
+        // so that it is not written to the database in the snapshot source table
```
we don't want to write it to the database because it is a transient asset that does not exist on the dataset?

snf2ye replied:
That's correct. We aren't ever writing this asset to the database so if we wrote an id in the snapshot source table that points to an id, it wouldn't be meaningful.

But, to Phil's point, I'm never populating this id field, so I removed this call.

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+    assetModel.rootTable("person");
+    assetModel.rootColumn("person_id");
+    // Manually add dictionary tables, leave columns empty to return all columns
+    assetModel.addTablesItem(new AssetTableModel().name("person"));
```
even though its the root table, it needs to be added again?

snf2ye replied:
That's correct - a table must be defined both as the root table and in the list of tables.

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +649,157 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+        table -> {
+          assetModel.addTablesItem(
+              new AssetTableModel().name(table.getDatasetTableName()).columns(table.getColumns()));
+          assetModel.follow(table.getRelationships());
```
what is the effect of adding the relationships? Does this include extra tables in output snapshots?

snf2ye replied:
Without relationships, all tables (besides the root table) would be empty. We can think of an asset as the instruction manual for creating a snapshot. And the relationships are the instructions for how to join between the root table and other tables (and then between other non-root tables) included in the spec when selecting which rows to include in the snapshot.

rjohanek commented on src/test/java/bio/terra/service/snapshotbuilder/SnapshotBuilderTestData.java:
```diff
@@ -155,21 +170,11 @@ private static SnapshotBuilderProgramDataListOption generateSnapshotBuilderProgr
-                  new SnapshotBuilderFeatureValueGroup()
-                      .name("Person")
-                      .values(List.of("Demographics Column 1", "Demographics Column 2"))))
+                  new SnapshotBuilderFeatureValueGroup().name("Condition"),
```
are these changes to the test data so that it lines up with the upcoming changes to the settings

snf2ye replied:
With this change, we expect for all columns listed in the request to map to real columns in the dataset. So, I'm removing these placeholder columns so the asset/snapshot creation does not break with this change. An alternate would be to populate this list with the real column names.

rjohanek commented on src/test/java/bio/terra/service/snapshotbuilder/SnapshotBuilderTestData.java:
```diff
@@ -179,24 +184,97 @@ private static SnapshotBuilderProgramDataListOption generateSnapshotBuilderProgr
+          .name(DrugExposure.TABLE_NAME)
+          .columns(
+              List.of(
+                  PERSON_ID_COLUMN,
```
I thought that columns would be empty? Why does this need to be populated?

snf2ye replied:
The columns will be empty in the asset and access request. 
However, this chunk of code is populating the dataset itself. Since we have relationships between columns and a root column in asset definition, it would complain if the dataset itself didn't have any columns.

snf2ye committed:
fetch asset instead of putting in flight map; Code review changes

Update CreateSnapshotByRequestIdInterfaceTest.java

cleanup

code review

snf2ye committed:
Preserve ordering for relationships

snf2ye committed:
Make sure both tables' relationships are added to the asset

update test

s-rubenstein commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,28 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
         Query query = Query.parse(snapshotQuery);
         String datasetName = query.getDatasetName();
-        createSnapshotTablesFromDatasetAsset(datasetName, assetName, snapshotSource, snapshot);
+        Dataset queryDataset = datasetService.retrieveByName(datasetName);
```
Is this not the same as `dataset` above? What is the difference between `datasetName` and `requestContents.getDatasetName`? And with my PR, should this change to using the datasetId that gets passed?

snf2ye replied:
Hm, this is a slightly different case and I'm little nervous about pulling this. Here we're grabbing the dataset form the parsed query. Which, really should be the same dataset and it would probably be bad if it was a different dataset. I think I would want to put this into a separate PR and put it in front of the DCJ team.

pshapiro4broad replied:
That's OK with me although I'd also want to factor out other data sources at the same time.

snf2ye replied:
I've added a ticket for us to revisit this in a separate PR. https://broadworkbench.atlassian.net/browse/DC-1167

rjohanek commented on src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotByRequestIdInterface.java:
```diff
@@ -18,26 +19,26 @@ default StepResult prepareAndCreateSnapshot(
-    AssetSpecification assetSpecification =
-        workingMap.get(
-            SnapshotWorkingMapKeys.SNAPSHOT_ASSET_SPECIFICATION, AssetSpecification.class);
+    AssetSpecification queryAssetSpecification =
```
I remember you bringing up some concerns about regenerating the asset, such as if the settings change. Is there anything we need to do to mitigate those concerns?

snf2ye replied:
That's a great question. I think this is something we can think about in the future -- If we follow similar patterns to elsewhere in TDR, we could add a "locking" mechanism for the source snapshot so we don't update the settings during request snapshot creation.

snf2ye replied:
I've added a post-MVP ticket: https://broadworkbench.atlassian.net/browse/DC-1155

pshapiro4broad replied:
In practice I think we should think of settings as being only set on snapshot creation. Right now we're still iterating on what the settings are, but if a snapshot's settings can change over time it could render a previous request invalid as it could change table mappings, etc. At some point in the future we remove the ability to set it later and only allow it to be set on snapshot creation.

s-rubenstein commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,28 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
-        String datasetName = snapshotRequestModel.getContents().get(0).getDatasetName();
-        createSnapshotTablesFromDatasetAsset(datasetName, ASSET_NAME, snapshotSource, snapshot);
+        String datasetName = requestContents.getDatasetName();
+        Dataset queryDataset = datasetService.retrieveByName(datasetName);
```
Same question - is this different from `dataset` above?

snf2ye replied:
Yes, this appears to be the same dataset. I'll remove this chunk of code.

snf2ye replied:
And I'll switch to grabbing the dataset via id instead of name.

s-rubenstein commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +645,154 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
-    AssetSpecification queryAssetSpecification = getAssetByNameFromDataset(queryDataset, assetName);
-    snapshotSource.assetSpecification(queryAssetSpecification);
-    conjureSnapshotTablesFromAsset(snapshot, snapshotSource);
+  public SnapshotAccessRequestResponse getSnapshotAccessRequestById(UUID accessRequestId) {
```
Does this need a helper method? I would say that `snapshotRequestDao.getById` would be clearer to me.

snf2ye replied:
getSnapshotAccessRequestById is also used in a couple of other places outside of the service. In those spots, I do this it is a good practice to call the service rather than directly calling the dao.

s-rubenstein commented on src/test/java/bio/terra/service/snapshot/flight/create/CreateSnapshotByRequestIdInterfaceTest.java:
```diff
@@ -69,15 +71,16 @@ void prepareAndCreateSnapshot() throws InterruptedException {
     // mock out call to createSnapshotPrimaryData
     when(createSnapshotByRequestIdInterface.createSnapshot(
-            eq(flightContext), eq(asset), eq(snapshot), eq(sqlQuery), any()))
+            any(), any(), eq(snapshot), eq(sqlQuery), any()))
```
Is it no longer possible to derive `flightContext` or `asset`?

snf2ye replied:
I added back flightContext, but the asset is now generated in the call so I don't think I can set it here.

snf2ye committed:
Code review changes

snf2ye committed:
Fix unit test

rjohanek reviewed (commented):
Do the integration test checks need to be updated to check for the right inclusion of tables, or it's already okay as is?

snf2ye commented:
> Do the integration test checks need to be updated to check for the right inclusion of tables, or it's already okay as is?

@rjohanek 
It should be okay as is! The asset should be built to match the previously hard-coded asset.

snf2ye committed:
Rework relationship walking

snf2ye committed:
first pass at updating column order

snf2ye committed:
Bug: follow direction not getting passed through to asset spec

snf2ye committed:
Revert "first pass at updating column order"

This reverts commit a5968fbc35bd8b882811feb1da950e55c5d4ab08.

snf2ye commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +644,155 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+        table -> {
+          assetModel.addTablesItem(
+              new AssetTableModel().name(table.getDatasetTableName()).columns(table.getColumns()));
+          // First add all person <-> occurrence relationships
+          assetModel.addFollowItem(table.getPersonRelationship());
+        });
+    // Second, add all occurrence <-> concept relationships
+    tables.forEach(table -> table.getConceptRelationships().forEach(assetModel::addFollowItem));
```
Note for reviewers: With the change to require strict relationships walking, I don't think we _need_ to break this into  person and then concept relationships. However, it might make it easier to correctly define the relationships in the snapshot builder settings for data custodians: It is clear that you need a person -> occurrence relationship and then occurrence -> concept relationship.

rjohanek commented:
> ### Remaining/Follow-on Changes
> * [ ]  Manually update snapshot builder settings per environment: After an initial review of these changes, I'm going to update snapshot builder settings across our environments to match the example in this code: Removing the placeholder column names and the genomic domain (since it doesn't map to a table in the current dataset).

Do we want to keep the genomic domain? Although we don't currently have genomic data I think it maps to the samples table

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +644,155 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+          assetModel.addTablesItem(
+              new AssetTableModel().name(table.getDatasetTableName()).columns(table.getColumns()));
+          // First add all person <-> occurrence relationships
+          assetModel.addFollowItem(table.getPersonRelationship());
```
so it is always the case that there is only ever one relationship to the person table for each occurrence table?

snf2ye replied:
Yeah, I think that should be the case.

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +644,155 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+                List.of("fpk_person_ethnicity_concept", "fpk_person_ethnicity_concept_s")));
+    tableMap.put(
+        "Race",
+        new SnapshotBuilderTable()
```
Why are year of birth, ethnicity, and race in the tableMap?

snf2ye replied:
Good point -I can remove them. These relationships will now be covered by the demographic option.

rjohanek commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +644,155 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+                    "fpk_observation_unit",
+                    "fpk_observation_qualifier",
+                    "fpk_observation_type_concept",
+                    "fpk_observation_value")));
```
so once this information is in the settings, the custodian will need to define this information twice is that right? Once on the dataset for all tables, and then again on the snapshot settings for each table. If that's the case it makes me want to revisit the auto generating settings once this upcoming settings work is done.

snf2ye replied:
Yes, that's correct. I think autogenerating the relationships would make a lot of sense for future improvement.

snf2ye replied:
Post-MVP ticket: https://broadworkbench.atlassian.net/browse/DC-1168

rjohanek commented on src/test/java/bio/terra/common/fixtures/DaoOperations.java:
```diff
@@ -105,7 +106,7 @@ public Snapshot ingestSnapshot(Snapshot snapshot) {
     SnapshotRequestModel snapshotRequest = createSnapshotRequestFromDataset(dataset, snapshotPath);
     Snapshot snapshot =
-        createSnapshotFromSnapshotRequest(snapshotRequest, dataset.getProjectResourceId());
+        createSnapshotFromSnapshotRequest(snapshotRequest, dataset, dataset.getProjectResourceId());
```
probably doesn't really matter, but if we are passing in the dataset now, we could just get the project resource Id from the dataset in the method instead of passing it in as well

snf2ye replied:
Good point!

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/AssetSpecification.java:
```diff
@@ -73,4 +74,13 @@ public AssetTable getAssetTableByName(String tableName) {
+    return followStrictDirection;
+  }
+
+  public AssetSpecification setFollowStrictDirection(boolean followStrictDirection) {
```
Since this a fluent-style method the convention is to not use `set`
```suggestion
  public AssetSpecification followStrictDirection(boolean followStrictDirection) {
```

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -331,12 +334,13 @@ private static List<AssetTable> processAssetTables(
-            .filter(map -> assetRelationshipNames.contains(map.getKey()))
-            .map(entry -> new AssetRelationship().datasetRelationship(entry.getValue()))
-            .collect(Collectors.toList()));
+      List<String> orderedAssetRelationshipNames, Map<String, Relationship> relationships) {
```
Does the name `assetRelationshipNames` need to be changed? A `List` is an ordered collection so this seems redundant to me.

pshapiro4broad commented on src/main/java/bio/terra/service/dataset/DatasetJsonConversion.java:
```diff
@@ -331,12 +334,13 @@ private static List<AssetTable> processAssetTables(
+        .filter(relationships::containsKey)
+        .map(
+            relationship ->
+                new AssetRelationship().datasetRelationship(relationships.get(relationship)))
```
Since the lambda type is `String` (and not `Relationship`) I would call this `name`
```suggestion
            name ->
                new AssetRelationship().datasetRelationship(relationships.get(name)))
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -579,7 +581,8 @@ public Snapshot retrieveByName(String name) {
    */
-  public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotRequestModel) {
+  public Snapshot makeSnapshotFromSnapshotRequest(
+      SnapshotRequestModel snapshotRequestModel, Dataset sourceDataset) {
```
I worry about the state duplication here, as the request model also has a dataset. Having both means that it's easier to make a mistake and use the "wrong" dataset by accident.

If we did want to address this, one way is to create a new type that includes the request data without the snapshot data sources. That way there's no way for a code to accidentally pick the wrong one.

snf2ye replied:
(As discussed in standup) I think we want to consider moving back to retrieving the dataset in this method as we were doing before - https://broadworkbench.atlassian.net/browse/DC-1166

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -607,15 +606,26 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+            buildAssetFromSnapshotAccessRequest(sourceDataset, accessRequest);
+        // populate the assetSpecification field so that we can write it to the working map in the
+        // step
+        snapshotSource.assetSpecification(queryAssetSpecification);
```
I think it's clearer to use a `set` API here instead of the fluent API. Since `SnapshotSource` is a class we control (not openapi generated) then we can have both types of setters.

snf2ye replied:
Wouldn't that depart for other calls in the SnapshotSource?

snf2ye replied:
I'll add the additional setAssetSpecification call here.

snf2ye replied:
Change ✅

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -589,13 +592,9 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
     }
 
     SnapshotRequestContentsModel requestContents = requestContentsList.get(0);
-    Dataset dataset = datasetService.retrieveByName(requestContents.getDatasetName());
```
It's a minor thing, but you could also call the parameter `dataset` since this local no longer exists, and the new parameter is the same concept as the old `dataset` local variable.

snf2ye replied:
Since we're going to revisit this change, I'm not going to rework the naming of these variables.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/SnapshotService.java:
```diff
@@ -633,12 +644,155 @@ public Snapshot makeSnapshotFromSnapshotRequest(SnapshotRequestModel snapshotReq
+    AssetModel assetModel =
+        new AssetModel()
+            .name("snapshot-by-request-asset")
+            .rootTable("person")
```
We should have a ticket in the backlog to replace these hardcoded values with values from `SnapshotBuilderSettings`. There are places in the query code that also assume the `person` table and the `person_id` key which should also be changed.

snf2ye replied:
I was thinking of potentially including at least setting the root table and column in https://broadworkbench.atlassian.net/browse/DC-1120.

pshapiro4broad commented on src/main/java/bio/terra/service/tabulardata/WalkRelationship.java:
```diff
@@ -27,21 +36,27 @@ public enum WalkDirection {
         .toColumn(datasetRelationship.getToColumn())
-        .toTable(datasetRelationship.getToTable());
+        .toTable(datasetRelationship.getToTable())
+        .setStrictRelationshipDirection(strictRelationshipDirection);
```
Did you consider making strict an aspect of the relationship itself? I guess since we want to enforce it for all relationships in an asset it would be a bit redundant, but maybe that's conceptually were it belongs.

snf2ye replied:
(we discussed this in person) But, I think that would be a good long term solution. For now, we're not propagating this parameter up to the user so we can test out its usefulness.

pshapiro4broad reviewed (commented):
comments so far

snf2ye committed:
Code review changes

- Don't pass in project id b/c can be retrieved from the dataset
- Include relationships for demographics
- don't include year of birth, ethnicity, or race

pshapiro4broad reviewed (approved):
Looks OK although I didn't read every line of test data

snf2ye committed:
code review: rename

snf2ye committed:
remove followStrictRelationship from asset because we're not storing it in the dataset

snf2ye committed:
Rename to more generic relationship names

snf2ye committed:
rename assetSpecification to setAssetSpecification
